gptfixer/gpt.c

```diff
@@ -0,0 +1,914 @@
+
+struct GPTPartitionEntry {
+  struct Guid PartitionTypeGUID;   /*   0 */
+  struct Guid UniquePartitionGUID; /*  16 */
+  uint64_t StartingLBA;            /*  32 */
+  uint64_t EndingLBA;              /*  40 */
+  uint64_t Attributes;             /*  48 */
+  uint8_t PartitionName[72];       /*  56 */
+};
+static_assert(sizeof(struct GPTPartitionEntry) == 128, "bug");
+
+struct __attribute__((packed, aligned(1))) MbrPartitionRecord {
+  uint8_t boot_indicator;
+  struct ChsSpec starting_chs;
+  uint8_t os_type;
+  struct ChsSpec ending_chs;
+  uint8_t starting_lba[4];
+  uint8_t size_in_lba[4];
+};
+#define ASSERT_OFFSET(a, b, offset, size)                                      \
+  static_assert(offsetof(a, b) == offset, "wrong offset");                     \
+  static_assert(sizeof(((a){}).b) == size, "wrong size");
+ASSERT_OFFSET(struct MbrPartitionRecord, boot_indicator, 0, 1);
+ASSERT_OFFSET(struct MbrPartitionRecord, starting_chs, 1, 3);
+ASSERT_OFFSET(struct MbrPartitionRecord, os_type, 4, 1);
+ASSERT_OFFSET(struct MbrPartitionRecord, ending_chs, 5, 3);
+ASSERT_OFFSET(struct MbrPartitionRecord, starting_lba, 8, 4);
+ASSERT_OFFSET(struct MbrPartitionRecord, size_in_lba, 12, 4);
+
+struct Mbr {
+  uint8_t boot_code[440];
+  uint8_t unique_mbr_disk_signature[4];
+  uint8_t unknown[2];
+  struct MbrPartitionRecord partition_record[4];
+  uint8_t signature[2];
+};
+ASSERT_OFFSET(struct Mbr, boot_code, 0, 440);
+ASSERT_OFFSET(struct Mbr, unique_mbr_disk_signature, 440, 4);
+ASSERT_OFFSET(struct Mbr, unknown, 444, 2);
+ASSERT_OFFSET(struct Mbr, partition_record, 446, 64);
+ASSERT_OFFSET(struct Mbr, signature, 510, 2);
+
+struct UUID {
+  uint8_t data[16];
+};
+
+// 512B -> 4096B:
+//
+// - Bytes 512 through 4096 will NOT be zero.
+//
+// 4096B -> 512B:
+//
+// - Bytes 512 through 4096 WILL be zero.
+
+// Step 1: Validate GPT
+// Step 2:
+
+bool bool_ioctl(int fd, long unsigned int op, void *arg) {
+  switch (ioctl(fd, op, arg)) {
+  case 0:
+    return true;
+  case -1:
+    return false;
+  default:
+    warnx("invalid return value from ioctl()");
+    abort();
+  }
+}
+
+[[gnu::format(printf, 1, 2)]]
+static void vlog(const char *fmt, ...) {
+  va_list args;
+  va_start(args, fmt);
+  if (fprintf(stderr, "fixgpt: ") != sizeof("fixgpt:") ||
+      vfprintf(stderr, fmt, args) < 0 ||
+      fputc((int)(unsigned char)'\n', stderr) != (int)(unsigned char)'\n')
+    err(2, "fixgpt: log write error");
+  va_end(args);
+}
+
+static bool check_mbr(struct Mbr *mbr, uint64_t size_in_sectors) {
+  bool changed = false;
+  if (mbr->signature[0] != 0x55 || mbr->signature[1] != 0xAA) {
+    errx(1,
+         "Invalid MBR signature: expected 0x55 0xAA, got 0x%02" PRIX8
+         " 0x%02" PRIX8,
+         mbr->signature[0], mbr->signature[1]);
+  }
+  if (mbr->partition_record[0].os_type != 0xEE) {
+    errx(1, "Bad protective MBR: OS type is not GPT Protective");
+  }
+  struct MbrPartitionRecord first_partition = {0};
+  for (int i = 2; i <= 4; ++i) {
```

Comment:
You know you can use `i+1` when formatting the message, right?
Right now you do `i-1` every time just to avoid doing `i+1` in case of unlikely error.
I know compiler will optimize this anyway, but the current version is at least unintuitive.